File: statsig-lib/src/statsig_user.rs
use crate::dyn_value;
use crate::evaluation::dynamic_value::DynamicValue;
use serde::{Deserialize, Serialize};
use std::collections::HashMap;

pub struct StatsigUserBuilder {
    pub user_id: Option<DynamicValue>,
    pub email: Option<DynamicValue>,
    pub ip: Option<DynamicValue>,
    pub user_agent: Option<DynamicValue>,
    pub country: Option<DynamicValue>,
    pub locale: Option<DynamicValue>,
    pub app_version: Option<DynamicValue>,
    pub custom: Option<HashMap<String, DynamicValue>>,
    pub private_attributes: Option<HashMap<String, DynamicValue>>,
    pub custom_ids: Option<HashMap<String, DynamicValue>>,
}

impl StatsigUserBuilder {
    pub fn new_with_user_id(user_id: String) -> Self {
        Self {
            user_id: Some(dyn_value!(user_id)),
            ..Self::new()
        }
    }

    pub fn new_with_custom_ids(custom_ids: HashMap<String, String>) -> Self {
        Self {
            custom_ids: Some(convert_str_map_to_dyn_values(custom_ids)),
            ..Self::new()
        }
    }

    fn new() -> Self {
        Self {
            user_id: None,
            email: None,
            ip: None,
            user_agent: None,
            country: None,
            locale: None,
            app_version: None,
            custom: None,
            private_attributes: None,
            custom_ids: None,
        }
    }

    pub fn user_id(mut self, user_id: Option<String>) -> Self {
        if let Some(user_id) = user_id {
            self.user_id = Some(dyn_value!(user_id));
        }
        self
    }

    pub fn custom_ids(mut self, custom_ids: Option<HashMap<String, String>>) -> Self {
        if let Some(custom_ids) = custom_ids {
            self.custom_ids = Some(convert_str_map_to_dyn_values(custom_ids));
        }
        self
    }

    pub fn email(mut self, email: Option<String>) -> Self {
        if let Some(email) = email {
            self.email = Some(dyn_value!(email));
        }
        self
    }

    pub fn ip(mut self, ip: Option<String>) -> Self {
        if let Some(ip) = ip {
            self.ip = Some(dyn_value!(ip));
        }
        self
    }

    pub fn user_agent(mut self, user_agent: Option<String>) -> Self {
        if let Some(user_agent) = user_agent {
            self.user_agent = Some(dyn_value!(user_agent));
        }
        self
    }

    pub fn country(mut self, country: Option<String>) -> Self {
        if let Some(country) = country {
            self.country = Some(dyn_value!(country));
        }
        self
    }

    pub fn locale(mut self, locale: Option<String>) -> Self {
        if let Some(locale) = locale {
            self.locale = Some(dyn_value!(locale));
        }
        self
    }

    pub fn app_version(mut self, app_version: Option<String>) -> Self {
        if let Some(app_version) = app_version {
            self.app_version = Some(dyn_value!(app_version));
        }
        self
    }

    // todo: support HashMap<String, String | Number | Boolean | Array<String>>
    pub fn custom_from_str_map(mut self, custom: Option<HashMap<String, String>>) -> Self {
        if let Some(custom) = custom {
            self.custom = Some(convert_str_map_to_dyn_values(custom));
        }
        self
    }

    pub fn custom(mut self, custom: Option<HashMap<String, DynamicValue>>) -> Self {
        if let Some(custom) = custom {
            self.custom = Some(custom);
        }
        self
    }

    // todo: support HashMap<String, String | Number | Boolean | Array<String>>
    pub fn private_attributes_from_str_map(
        mut self,
        private_attributes: Option<HashMap<String, String>>,
    ) -> Self {
        if let Some(private_attributes) = private_attributes {
            self.private_attributes = Some(convert_str_map_to_dyn_values(private_attributes));
        }
        self
    }

    pub fn private_attributes(
        mut self,
        private_attributes: Option<HashMap<String, DynamicValue>>,
    ) -> Self {
        if let Some(private_attributes) = private_attributes {
            self.private_attributes = Some(private_attributes);
        }
        self
    }

    pub fn build(self) -> StatsigUser {
        StatsigUser {
            user_id: self.user_id,
            email: self.email,
            ip: self.ip,
            user_agent: self.user_agent,
            country: self.country,
            locale: self.locale,
            app_version: self.app_version,
            custom: self.custom,
            private_attributes: self.private_attributes,
            custom_ids: self.custom_ids,
        }
    }
}

#[derive(Clone, Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct StatsigUser {
    #[serde(rename = "userID", skip_serializing_if = "Option::is_none")]
    pub user_id: Option<DynamicValue>,

    #[serde(skip_serializing_if = "Option::is_none")]
    pub email: Option<DynamicValue>,

    #[serde(skip_serializing_if = "Option::is_none")]
    pub ip: Option<DynamicValue>,

    #[serde(skip_serializing_if = "Option::is_none")]
    pub user_agent: Option<DynamicValue>,

    #[serde(skip_serializing_if = "Option::is_none")]
    pub country: Option<DynamicValue>,

    #[serde(skip_serializing_if = "Option::is_none")]
    pub locale: Option<DynamicValue>,

    #[serde(skip_serializing_if = "Option::is_none")]
    pub app_version: Option<DynamicValue>,

    #[serde(skip_serializing_if = "Option::is_none")]
    pub custom: Option<HashMap<String, DynamicValue>>,

    #[serde(skip_serializing_if = "Option::is_none")]
    pub private_attributes: Option<HashMap<String, DynamicValue>>,

    #[serde(rename = "customIDs", skip_serializing_if = "Option::is_none")]
    pub custom_ids: Option<HashMap<String, DynamicValue>>,
}

impl StatsigUser {
    pub fn with_user_id(user_id: String) -> Self {
        StatsigUser {
            user_id: Some(dyn_value!(user_id)),
            ..Self::default()
        }
    }

    pub fn with_custom_ids(custom_ids: HashMap<String, String>) -> Self {
        StatsigUser {
            custom_ids: Some(convert_str_map_to_dyn_values(custom_ids)),
            ..Self::default()
        }
    }

    fn default() -> Self {
        StatsigUser {
            user_id: None,
            email: None,
            ip: None,
            user_agent: None,
            country: None,
            locale: None,
            app_version: None,
            custom: None,
            private_attributes: None,
            custom_ids: None,
        }
    }
}

fn convert_str_map_to_dyn_values(
    custom_ids: HashMap<String, String>,
) -> HashMap<String, DynamicValue> {
    custom_ids
        .into_iter()
        .map(|(k, v)| (k, dyn_value!(v)))
        .collect()
}

File: statsig-go/statsig_user.go
package statsig_go_core

import (
	"encoding/json"
	"fmt"
	"runtime"
)

type StatsigUser struct {
	ref uint64
}

// todo: introduce custom type for handling valid JSON primitives only instead of using 'any'
type StatsigUserBuilder struct {
	UserID string `json:"userID"`
	// map[string] string | number
	CustomIDs  map[string]any `json:"customIDs"`
	Email      *string        `json:"email"`
	IpAddress  *string        `json:"ip"`
	UserAgent  *string        `json:"userAgent"`
	Country    *string        `json:"country"`
	Locale     *string        `json:"locale"`
	AppVersion *string        `json:"appVersion"`
	// map[string] string | number | boolean | array<string>
	Custom *map[string]any `json:"custom"`
	// map[string] string | number | boolean | array<string>
	PrivateAttributes *map[string]any `json:"privateAttributes"`

	// nilUserID, set via WithoutUserID, serializes the user with no userID field
	// at all (a true null/absent userID at the FFI boundary, distinct from an
	// empty-string userID). When false, Build serializes UserID as-is ("" when
	// unset) — the default, backward-compatible wire shape.
	nilUserID bool
}

func NewUserBuilderWithUserID(userID string) *StatsigUserBuilder {
	return &StatsigUserBuilder{
		UserID: userID,
	}
}

func NewUserBuilderWithCustomIDs(customIDs map[string]any) *StatsigUserBuilder {
	return &StatsigUserBuilder{
		CustomIDs: customIDs,
	}
}

func (b *StatsigUserBuilder) WithUserID(userID string) *StatsigUserBuilder {
	b.UserID = userID
	b.nilUserID = false
	return b
}

// WithoutUserID builds a user with no userID: the serialized payload omits the
// userID field entirely, representing a true null/absent userID to statsig-rust
// (distinct from an empty-string userID). Use this for users identified solely by
// custom IDs.
//
// This is a short-term, non-breaking addition. The long-term direction is to type
// UserID as *string so null is representable directly; WithoutUserID will be
// retired in favor of a nil UserID when that lands.
func (b *StatsigUserBuilder) WithoutUserID() *StatsigUserBuilder {
	// nilUserID alone drives omission in marshalUserJSON; UserID is left untouched.
	b.nilUserID = true
	return b
}

func (b *StatsigUserBuilder) WithCustomIDs(customIDs map[string]any) *StatsigUserBuilder {
	b.CustomIDs = customIDs
	return b
}

func (b *StatsigUserBuilder) WithEmail(email string) *StatsigUserBuilder {
	b.Email = &email
	return b
}

func (b *StatsigUserBuilder) WithIpAddress(ipAddress string) *StatsigUserBuilder {
	b.IpAddress = &ipAddress
	return b
}

func (b *StatsigUserBuilder) WithUserAgent(userAgent string) *StatsigUserBuilder {
	b.UserAgent = &userAgent
	return b
}

func (b *StatsigUserBuilder) WithCountry(country string) *StatsigUserBuilder {
	b.Country = &country
	return b
}

func (b *StatsigUserBuilder) WithLocale(locale string) *StatsigUserBuilder {
	b.Locale = &locale
	return b
}

func (b *StatsigUserBuilder) WithAppVersion(appVersion string) *StatsigUserBuilder {
	b.AppVersion = &appVersion
	return b
}

func (b *StatsigUserBuilder) WithCustom(custom map[string]any) *StatsigUserBuilder {
	b.Custom = &custom
	return b
}

func (b *StatsigUserBuilder) WithPrivateAttributes(privateAttributes map[string]any) *StatsigUserBuilder {
	b.PrivateAttributes = &privateAttributes
	return b
}

// marshalUserJSON renders the wire JSON sent to statsig-rust without mutating the
// builder. By default an unset UserID serializes as "" (the legacy wire shape);
// when WithoutUserID was called, the userID field is omitted entirely.
//
// The alias type drops StatsigUserBuilder's methods while keeping its json tags;
// the outer UserID *string shadows alias.UserID (a shallower field wins in
// encoding/json), giving a single userID key with omit-on-nil behavior. All other
// fields serialize from the embedded value, so new builder fields need no change
// here.
func (b *StatsigUserBuilder) marshalUserJSON() ([]byte, error) {
	type alias StatsigUserBuilder
	wire := struct {
		*alias
		UserID *string `json:"userID,omitempty"`
	}{alias: (*alias)(b)}

	if !b.nilUserID {
		wire.UserID = &b.UserID
	}

	return json.Marshal(wire)
}

func (b *StatsigUserBuilder) Build() (*StatsigUser, error) {
	jsonData, err := b.marshalUserJSON()
	if err != nil {
		return nil, fmt.Errorf("error marshalling user: %v", err)
	}

	userRef := GetFFI().statsig_user_create_from_data(
		string(jsonData),
	)

	if userRef == 0 {
		return nil, fmt.Errorf("error creating StatsigUser")
	}

	user := &StatsigUser{
		ref: userRef,
	}

	runtime.SetFinalizer(user, func(obj *StatsigUser) {
		GetFFI().statsig_user_release(obj.ref)
	})

	return user, nil
}
